pkg/mkverification-raw-efi/vmcap.go

```diff
@@ -0,0 +1,373 @@
+// Copyright (c) 2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+)
+
+// For more information chekout out https://man7.org/linux/man-pages/man4/msr.4.html
+// and https://cdrdv2.intel.com/v1/dl/getContent/671200 (Appending A -VMX capability
+// reporting facility, on page 4483)
+
+const (
+	basic             = 0x480
+	pinBasedCtls      = 0x481
+	procBasedCtls     = 0x482
+	exitCtls          = 0x483
+	entryCtls         = 0x484
+	miscCtls          = 0x485
+	procBasedCtls2    = 0x48B
+	eptVPIDCap        = 0x48C
+	truePinBasedCtls  = 0x48D
+	trueProcBasedCtls = 0x48E
+	trueExitCtls      = 0x48F
+	trueEntryCtls     = 0x490
+	vmFunc            = 0x491
+	procBasedCtls3    = 0x492
+)
+
+type msr struct {
+	f *os.File
+}
+
+func newMsr() (*msr, error) {
+	f, err := os.OpenFile("/dev/cpu/0/msr", os.O_RDONLY, 0)
+	if err != nil {
+		f, err = os.OpenFile("/dev/msr0", os.O_RDONLY, 0)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &msr{f: f}, nil
+}
+
+func (m *msr) read(index uint64, defaultVal uint64) (uint64, error) {
+	p := make([]byte, 8)
+	_, err := m.f.Seek(int64(index), 0)
+	if err != nil {
+		return defaultVal, err
+	}
+	_, err = m.f.Read(p)
+	if err != nil {
+		return defaultVal, err
+	}
+
+	data := binary.LittleEndian.Uint64(p)
+
+	return data, nil
+}
+
+type control struct {
+	name       string
+	bits       map[int]string
+	msr     uint64
+	trueMsr uint64
+}
+
+func (c *control) read(nr uint64) (uint32, uint32) {
+	m, _ := newMsr()
+	val, _ := m.read(nr, 0)
+	return uint32(val & 0xffffffff), uint32(val >> 32)
+}
+
+func (c *control) show() {
+	fmt.Println(c.name)
+	mb1, cb1 := c.read(c.msr)
+	tmb1, tcb1 := uint32(0), uint32(0)
+	if c.trueMsr != 0 {
+		tmb1, tcb1 = c.read(c.trueMsr)
+	}
+	for bit := range c.bits {
+		zero := !(mb1&(1<<bit) != 0)
+		one := cb1&(1<<bit) != 0
+		trueZero := !(tmb1&(1<<bit) != 0)
+		trueOne := tcb1&(1<<bit) != 0
+		s := "?"
+		if c.trueMsr != 0 && trueZero && trueOne && one && !zero {
+			s = "default"
+		} else if zero && !one {
+			s = "no"
+		} else if one && !zero {
+			s = "forced"
+		} else if one && zero {
+			s = "yes"
+		}
+		fmt.Printf("  %-40s %s\n", c.bits[bit], s)
+	}
+}
+
+type allowedControl struct {
+	*control
+}
+
+func (a *allowedControl) read(nr uint64) (int, uint64) {
+	m, _ := newMsr()
+	val, _ := m.read(nr, 0)
```

Comment:
@jsfakian , sorry for the long delay... I've tested and it's working, I just would like to point out a detail here, if you ignore errors from newMsr() it can lead to the following segfault when newMsr() fail:

```
Basic VMX Information
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x480a61]

goroutine 1 [running]:
main.(*msr).read(0x0, 0xc000062010?, 0xc00009eaa0?)
	/tmp/vmcap.go:52 +0x21
main.(*misc).show(0xc00009ebe8)
	/tmp/vmcap.go:124 +0x8f
main.main()
	/tmp/vmcap.go:365 +0x220f
```

I think it should fail most likely when running without root, which will not be the case here. However, a manual execution on ARM will lead to the segfault anyways, even running with root, so I think would be good to catch this error instead of ignore it...